File: frontend/jwst-frontend/src/components/mast/map/healpix.ts
/**
 * HEALPix NESTED `ang2pix` — the same algorithm as the backend's
 * `app/mast/healpix.py`, so a click on the coverage grid can be matched to
 * the cell the server counted (Górski et al. 2005; healpy's C code).
 * Pinned values in the test come from astropy_healpix.
 */

const TWO_OVER_PI = 2 / Math.PI;

function spreadBits(v: number): bigint {
  let x = BigInt(v);
  x = (x | (x << 16n)) & 0x0000ffff0000ffffn;
  x = (x | (x << 8n)) & 0x00ff00ff00ff00ffn;
  x = (x | (x << 4n)) & 0x0f0f0f0f0f0f0f0fn;
  x = (x | (x << 2n)) & 0x3333333333333333n;
  x = (x | (x << 1n)) & 0x5555555555555555n;
  return x;
}

export function orderForNside(nside: number): number {
  if (!Number.isInteger(nside) || nside <= 0 || (nside & (nside - 1)) !== 0) {
    throw new Error(`nside must be a power of two, got ${nside}`);
  }
  return Math.round(Math.log2(nside));
}

/** NESTED pixel index of (RA, Dec) in degrees. */
export function ang2pixNest(nside: number, raDeg: number, decDeg: number): number {
  const order = orderForNside(nside);
  const ra = ((raDeg % 360) + 360) % 360;
  const dec = Math.max(-90, Math.min(90, decDeg));
  const phi = (ra * Math.PI) / 180;
  const z = Math.sin((dec * Math.PI) / 180);
  const za = Math.abs(z);
  const tt = (phi % (2 * Math.PI)) * TWO_OVER_PI; // [0, 4)

  let face: number;
  let ix: number;
  let iy: number;
  if (za <= 2 / 3) {
    const temp1 = nside * (0.5 + tt);
    const temp2 = nside * z * 0.75;
    const jp = Math.floor(temp1 - temp2);
    const jm = Math.floor(temp1 + temp2);
    const ifp = Math.floor(jp / nside);
    const ifm = Math.floor(jm / nside);
    face = ifp === ifm ? ifp | 4 : ifp < ifm ? ifp : ifm + 8;
    ix = jm & (nside - 1);
    iy = nside - (jp & (nside - 1)) - 1;
  } else {
    const ntt = Math.min(Math.floor(tt), 3);
    const tp = tt - ntt;
    const tmp = nside * Math.sqrt(3 * (1 - za));
    const jp = Math.min(Math.floor(tp * tmp), nside - 1);
    const jm = Math.min(Math.floor((1 - tp) * tmp), nside - 1);
    if (z >= 0) {
      face = ntt;
      ix = nside - jm - 1;
      iy = nside - jp - 1;
    } else {
      face = ntt + 8;
      ix = jp;
      iy = jm;
    }
  }
  const pix = (BigInt(face) << BigInt(2 * order)) + spreadBits(ix) + (spreadBits(iy) << 1n);
  return Number(pix);
}

File: processing-engine/app/mast/healpix.py
"""Minimal HEALPix (NESTED scheme) pixel indexing in pure numpy.

The coverage density grid (MAST Search v2 Phase 5) bins observation
centroids into HEALPix cells so the frontend can draw them as a MOC
(``A.MOCFromJSON`` takes NESTED indices per order). The full engine image
has ``astropy_healpix`` via reproject, but the lightweight MAST proxy image
(``requirements-mast.txt``) does not — and a dependency for one function is
not worth it. This is the standard ``ang2pix_nest`` algorithm (Górski et
al. 2005; healpy's C implementation), vectorised. ``tests/test_healpix.py``
cross-checks it against ``astropy_healpix`` when that is importable.
"""

from __future__ import annotations

import numpy as np


def _spread_bits(v: np.ndarray) -> np.ndarray:
    """Interleave zeros between the bits of ``v`` (uint64): b3 b2 b1 b0 → 0b3 0b2 0b1 0b0."""
    v = v.astype(np.uint64)
    v = (v | (v << np.uint64(16))) & np.uint64(0x0000FFFF0000FFFF)
    v = (v | (v << np.uint64(8))) & np.uint64(0x00FF00FF00FF00FF)
    v = (v | (v << np.uint64(4))) & np.uint64(0x0F0F0F0F0F0F0F0F)
    v = (v | (v << np.uint64(2))) & np.uint64(0x3333333333333333)
    v = (v | (v << np.uint64(1))) & np.uint64(0x5555555555555555)
    return v


def order_for_nside(nside: int) -> int:
    if nside <= 0 or (nside & (nside - 1)) != 0:
        raise ValueError(f"nside must be a power of two, got {nside}")
    return nside.bit_length() - 1


def ang2pix_nest(nside: int, ra_deg, dec_deg) -> np.ndarray:
    """NESTED pixel index for each (RA, Dec) in degrees. Vectorised; returns int64."""
    order_for_nside(nside)
    ra = np.atleast_1d(np.asarray(ra_deg, dtype=np.float64))
    dec = np.atleast_1d(np.asarray(dec_deg, dtype=np.float64))
    if ra.shape != dec.shape:
        raise ValueError("ra and dec must have the same shape")

    phi = np.deg2rad(np.mod(ra, 360.0))
    z = np.sin(np.deg2rad(np.clip(dec, -90.0, 90.0)))
    za = np.abs(z)
    tt = np.mod(phi, 2 * np.pi) * (2.0 / np.pi)  # in [0, 4)

    ns = float(nside)
    ix = np.empty(ra.shape, dtype=np.int64)
    iy = np.empty(ra.shape, dtype=np.int64)
    face = np.empty(ra.shape, dtype=np.int64)

    # Equatorial region: |z| <= 2/3
    eq = za <= 2.0 / 3.0
    if np.any(eq):
        temp1 = ns * (0.5 + tt[eq])
        temp2 = ns * z[eq] * 0.75
        jp = (temp1 - temp2).astype(np.int64)  # ascending edge line index
        jm = (temp1 + temp2).astype(np.int64)  # descending edge line index
        ifp = jp // nside
        ifm = jm // nside
        f = np.where(ifp == ifm, ifp | 4, np.where(ifp < ifm, ifp, ifm + 8))
        face[eq] = f
        ix[eq] = jm & (nside - 1)
        iy[eq] = nside - (jp & (nside - 1)) - 1

    # Polar caps: |z| > 2/3
    po = ~eq
    if np.any(po):
        ntt = np.minimum(tt[po].astype(np.int64), 3)
        tp = tt[po] - ntt
        tmp = ns * np.sqrt(3.0 * (1.0 - za[po]))
        jp = np.minimum((tp * tmp).astype(np.int64), nside - 1)
        jm = np.minimum(((1.0 - tp) * tmp).astype(np.int64), nside - 1)
        north = z[po] >= 0
        face[po] = np.where(north, ntt, ntt + 8)
        ix[po] = np.where(north, nside - jm - 1, jp)
        iy[po] = np.where(north, nside - jp - 1, jm)

    order = order_for_nside(nside)
    pix = (
        (face.astype(np.uint64) << np.uint64(2 * order))
        + _spread_bits(ix)
        + (_spread_bits(iy) << np.uint64(1))
    )
    return pix.astype(np.int64)
